File: typescript/src/fixedPriceLBP/fixedPriceLBP.ts
import { MathSol } from '../utils/math';
import { toRawUndoRateRoundDown } from '../vault/utils';
import {
    MaxSingleTokenRemoveParams,
    MaxSwapParams,
    type PoolBase,
    Rounding,
    SwapKind,
    type SwapParams,
} from '../vault/types';
import { MAX_UINT256 } from '../constants';

import type { FixedPriceLBPState } from './data';

export class FixedPriceLBP implements PoolBase {
    projectTokenIndex: number;
    reserveTokenIndex: number;
    projectTokenRate: bigint;
    isSwapEnabled: boolean;

    constructor(poolState: FixedPriceLBPState) {
        this.projectTokenIndex = poolState.projectTokenIndex;
        this.reserveTokenIndex = poolState.reserveTokenIndex;
        this.projectTokenRate = poolState.projectTokenRate;
        this.isSwapEnabled = poolState.isSwapEnabled;
    }

    getMaximumInvariantRatio(): bigint {
        return MAX_UINT256;
    }

    getMinimumInvariantRatio(): bigint {
        return 0n;
    }

    getMaxSwapAmount(maxSwapParams: MaxSwapParams): bigint {
        const {
            balancesLiveScaled18,
            indexIn,
            indexOut,
            tokenRates,
            scalingFactors,
            swapKind,
        } = maxSwapParams;

        if (swapKind === SwapKind.GivenIn) {
            // GivenIn: user specifies reserve amount in.
            // Max input is limited by output (project) token balance.
            // maxIn (scaled18) = projectBalance * projectTokenRate
            const maxIn18 = MathSol.mulDownFixed(
                balancesLiveScaled18[indexOut],
                this.projectTokenRate,
            );
            // Convert to raw using input token's scaling/rate
            return toRawUndoRateRoundDown(
                maxIn18,
                scalingFactors[indexIn],
                tokenRates[indexIn],
            );
        }

        // GivenOut: user specifies project amount out.
        // Max output is the project token balance.
        const maxOut18 = balancesLiveScaled18[indexOut];
        return toRawUndoRateRoundDown(
            maxOut18,
            scalingFactors[indexOut],
            tokenRates[indexOut],
        );
    }

    getMaxSingleTokenAddAmount(): bigint {
        throw new Error('UnsupportedOperation');
    }

    getMaxSingleTokenRemoveAmount(
        _maxRemoveParams: MaxSingleTokenRemoveParams,
    ): bigint {
        throw new Error('UnsupportedOperation');
    }

    onSwap(swapParams: SwapParams): bigint {
        if (!this.isSwapEnabled) {
            throw new Error('SwapsDisabled');
        }

        if (swapParams.indexIn === this.projectTokenIndex) {
            throw new Error('SwapOfProjectTokenIn');
        }

        if (swapParams.swapKind === SwapKind.GivenIn) {
            // Reserve tokens in, project tokens out: amountOut = amountIn / rate
            return MathSol.divDownFixed(
                swapParams.amountGivenScaled18,
                this.projectTokenRate,
            );
        }

        // ExactOut: amountIn = amountOut * rate
        return MathSol.mulUpFixed(
            swapParams.amountGivenScaled18,
            this.projectTokenRate,
        );
    }

    computeInvariant(
        balancesLiveScaled18: bigint[],
        rounding: Rounding,
    ): bigint {
        // inv = projectBalance * rate + reserveBalance
        const projectTokenValue =
            rounding === Rounding.ROUND_UP
                ? MathSol.mulUpFixed(
                      balancesLiveScaled18[this.projectTokenIndex],
                      this.projectTokenRate,
                  )
                : MathSol.mulDownFixed(
                      balancesLiveScaled18[this.projectTokenIndex],
                      this.projectTokenRate,
                  );

        return (
            projectTokenValue +
            balancesLiveScaled18[this.reserveTokenIndex]
        );
    }

    computeBalance(
        _balancesLiveScaled18: bigint[],
        _tokenInIndex: number,
        _invariantRatio: bigint,
    ): bigint {
        throw new Error('UnsupportedOperation');
    }
}

File: python/src/pools/fixed_price_lbp/fixed_price_lbp.py
from typing import List

from src.common.maths import Rounding, div_down_fixed, mul_down_fixed, mul_up_fixed
from src.common.pool_base import PoolBase
from src.common.swap_params import SwapParams
from src.common.types import SwapKind
from src.pools.fixed_price_lbp.fixed_price_lbp_data import FixedPriceLBPState

MAX_UINT256 = 2**256 - 1


class FixedPriceLBP(PoolBase):
    def __init__(self, pool_state: FixedPriceLBPState):
        self.project_token_index = pool_state.project_token_index
        self.reserve_token_index = pool_state.reserve_token_index
        self.project_token_rate = pool_state.project_token_rate
        self.is_swap_enabled = pool_state.is_swap_enabled

    def get_maximum_invariant_ratio(self) -> int:
        return MAX_UINT256

    def get_minimum_invariant_ratio(self) -> int:
        return 0

    def on_swap(self, swap_params: SwapParams) -> int:
        if not self.is_swap_enabled:
            raise ValueError("SwapsDisabled")

        if swap_params.index_in == self.project_token_index:
            raise ValueError("SwapOfProjectTokenIn")

        if swap_params.swap_kind.value == SwapKind.GIVENIN.value:
            # Reserve tokens in, project tokens out: amountOut = amountIn / rate
            return div_down_fixed(
                swap_params.amount_given_scaled18, self.project_token_rate
            )

        # ExactOut: amountIn = amountOut * rate
        return mul_up_fixed(swap_params.amount_given_scaled18, self.project_token_rate)

    def compute_invariant(
        self, balances_live_scaled18: List[int], rounding: Rounding
    ) -> int:
        # inv = projectBalance * rate + reserveBalance
        if rounding.value == Rounding.ROUND_UP.value:
            project_token_value = mul_up_fixed(
                balances_live_scaled18[self.project_token_index],
                self.project_token_rate,
            )
        else:
            project_token_value = mul_down_fixed(
                balances_live_scaled18[self.project_token_index],
                self.project_token_rate,
            )

        return project_token_value + balances_live_scaled18[self.reserve_token_index]

    def compute_balance(
        self,
        balances_live_scaled18: List[int],
        token_in_index: int,
        invariant_ratio: int,
    ) -> int:
        raise ValueError("UnsupportedOperation")
